### DISCUSSION TITLE
How to select a tree node and expand it

### DISCUSSION THREAD

**Discussion by davetapley** [Textualize/textual#2135] [Q&A] [answered]
I have a populated a `Tree` and collected the `NodeID`s  in dictionary which maps the items to their nodes.

How can I select that node and expand the tree to show it exactly as if the user had manually expanded clicked on it?

I see a few methods like [`select_node`](https://textual.textualize.io/api/tree/#textual.widgets._tree.Tree.select_node), but they don't seem to do that 🤔

**Comment by TomJGooding** [comment, marked as answer]
You might find the code references in the docs for [Tree](https://textual.textualize.io/api/tree/#textual.widgets.Tree) and  [TreeNode](https://textual.textualize.io/api/tree/#textual.widgets._tree.TreeNode) helpful.

`select_node` will only "Move the cursor to the given node". You can change the expanded state of a `TreeNode` using the `toggle` or `expand` methods.

EDIT: It looks like you could also use `action_toggle_node`, but only after ensuring that the selected node has refreshed: `tree.call_after_refresh(tree.action_toggle_node)`

```python
from textual.app import App, ComposeResult
from textual.widgets import Tree


class TreeApp(App):
    def compose(self) -> ComposeResult:
        tree: Tree[dict] = Tree("Dune")
        tree.root.expand()
        characters = tree.root.add("Characters")
        characters.add_leaf("Paul")
        characters.add_leaf("Jessica")
        characters.add_leaf("Chani")
        yield tree

        node_to_expand = tree.get_node_by_id(characters.id)
        tree.select_node(node_to_expand)
        node_to_expand.expand()


if __name__ == "__main__":
    app = TreeApp()
    app.run()
```

**Reply by TomJGooding** [reply]
For context as to why selecting a node is separate from expanding it, see [this issue](https://github.com/Textualize/textual/issues/1433).
